### PR TITLE
chore(ci): Generate circuit benchmark summaries on PRs

### LINKED ISSUE #3182: Benchmark number of ACIR opcodes per PR
### Problem

As surfaced in https://github.com/noir-lang/noir/issues/2720#issuecomment-1723507947, the CI does not currently check for changes in the number of ACVM opcodes generated by a set of benchmark Noir programs.

This leads to the situation where certain PRs could have bumped the number of ACVM opcodes (hence degrade compilation and proving speeds) of certain Noir keywords / language features, where core contributors:
- Would not have noticed until the community reports it
- Would have to go through a painful debugging process to identify the PR that caused the bump

### Happy Case

A set of benchmarking programs should be aligned upon, where the CI should then:
1. Automatically benchmark the number of ACIR opcodes generated by the programs on each PR
2. Highlight programs that came with considerable bumps (e.g. ≥+10%) in number of ACIR opcodes

### Alternatives Considered

_No response_

### Additional Context

A significant portion of the work could be borrowed from Aztec's implementation: https://github.com/AztecProtocol/aztec-packages/pull/2733

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #3182 

## Summary\*

This PR adds benchmarking for how PRs affect the number of opcodes/constraints generated for the programs in the `execution_success` test suite. If a PR changes the ACIR generated then a sticky comment will be added to the PR with a summary of the changes.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Here's an example output

---

# Changes to circuit sizes

> Generated at commit: [86cee4c3845bef99f38a493f19e40914a2a7ea71](/TomAFrench/test-noir-repo/commit/86cee4c3845bef99f38a493f19e40914a2a7ea71), compared to commit: [eed7b9913b29686799bb1f07b54a2722b8c361e5](/TomAFrench/test-noir-repo/commit/eed7b9913b29686799bb1f07b54a2722b8c361e5)

### 🧾 Summary (10% most significant diffs)

| Program | ACIR opcodes (+/-) | % | Circuit size (+/-) | % |
|:-|-:|-:|-:|-:|
| **test_program** | -10 ✅ | **-22.73%** | -10 ✅ | **-0.36%** |
---

<details>
<summary><strong>Full diff report</strong> 👇</summary>
<br />

| Program | ACIR opcodes (+/-) | % | Circuit size (+/-) | % |
|:-|-:|-:|-:|-:|
| **test_program** | 34&nbsp;(-10) | **-22.73%** | 2,778&nbsp;(-10) | **-0.36%** |
</details>

**Comment by kevaundray** [comment]
wow :D

**Comment by TomAFrench** [comment]
This may need an admin merge. Looks like the action is going to search through the entire history for a baseline benchmark artifact.

**Comment by TomAFrench** [comment]
We can improve the action to simplify the bootstrapping experience but this should allow us to get started using it.

File: .github/workflows/gates_report.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+      - name: Generate gates report
+        working-directory: ./tooling/nargo_cli/tests
+        run: |
+          ./gates_report.sh
+          mv gates_report.json ../../../gates_report.json
+       
+      - name: Compare gates reports
+        id: gates_diff
+        uses: TomAFrench/noir-gates-diff@e7cf131b7e7f044c01615f93f0b855f65ddc02d4
```
Comment (kevaundray) [inline .github/workflows/gates_report.yml:79]: we can change this in a followup PR

Comment (TomAFrench) [reply .github/workflows/gates_report.yml:79]: Yeah, we need to do some iterations before we can make a proper releases of the action.

**Pull request merged**